File: deckOfCards/suit.rb
class Suit
    attr_reader :value
  
    def initialize(value)
      @value = value
    end
  
    Club = Suit.new(0)
    Diamond = Suit.new(1)
    Heart = Suit.new(2)
    Spade = Suit.new(3)
  
    def self.getSuitFromValue(value)
      case value
      when 0 then Club
      when 1 then Diamond
      when 2 then Heart
      when 3 then Spade
      else nil
      end
    end
  end
  
  class Card
    attr_accessor :available
    attr_reader :faceValue, :suit
  
    def initialize(faceValue, suit)
      @available = true
      @faceValue = faceValue
      @suit = suit
    end
  
    def value
      raise NotImplementedError, "Subclasses must implement the 'value' method"
    end
  
    def suit
      @suit
    end
  
    def isAvailable
      @available
    end
  
    def markUnavailable
      @available = false
    end
  
    def markAvailable
      @available = true
    end
  end
  
  class Deck
    attr_accessor :cards, :dealtIndex
  
    def initialize
      @cards = []
      @dealtIndex = 0
    end
  
    def setDeckOfCards(deckOfCards)
      @cards = deckOfCards
    end
  
    def shuffle
      @cards.shuffle!
      @dealtIndex = 0
    end
  
    def remainingCards
      @cards.size - @dealtIndex
    end
  
    def dealHand(number)
      return nil if remainingCards < number
  
      hand = @cards[@dealtIndex, number]
      @dealtIndex += number
      hand
    end
  
    def dealCard
      return nil if remainingCards.zero?
  
      card = @cards[@dealtIndex]
      @dealtIndex += 1
      card
    end
  end
  
  class Hand
    attr_accessor :cards
  
    def initialize
      @cards = []
    end
  
    def score
      @cards.sum(&:value)
    end
  
    def addCard(card)
      @cards << card
    end
  end

File: deckOfCards/suit.py
from enum import Enum
from abc import ABC, abstractmethod
import random

class Suit(Enum):
    Club = 0
    Diamond = 1
    Heart = 2
    Spade = 3

    def __new__(cls, value):
        obj = object.__new__(cls)
        obj._value_ = value
        return obj

    @staticmethod
    def get_suit_from_value(value):
        return Suit(value) if 0<= value < 4 else None
    
class Card(ABC):
    def __init__(self, face_value, suit):
        self.available = True
        self.face_value = face_value
        self.suit = suit

    @abstractmethod
    def value(self):
        pass

    def is_available(self):
        return self.available

    def mark_unavailable(self):
        self.available = False

    def mark_available(self):
        self.available = True

class Deck(ABC):
    def __init__(self):
        self.cards = []
        self.dealt_index = 0

    def set_deck_of_cards(self, deck_of_cards):
        self.cards = deck_of_cards
    
    def shuffle(self):
        random.shuffle(self.cards)
        self.dealt_index = 0
    
    def remaining_cards(self):
        return len(self.cards) - self.dealt_index
    
    def deal_hand(self, number):
        if self.remaining_cards() >= number:
            hand = self.cards[self.dealt_index:self.dealt_index + number]
            self.dealt_index += number
            return hand
        else:
            return None
    
    def deal_card(self):
        if self.remaining_cards() > 0:
            card = self.cards[self.dealt_index]
            self.dealt_index += 1
            return card
        else:
            return None

class Hand(ABC):
    def __init__(self):
        self.cards = []

    def score(self):
        return sum(card.value() for card in self.cards)
    
    def add_card(self, card):
        self.cards.append(card)